### PR TITLE
Topic mining

### PR DESCRIPTION
**Pull request by AlbertCerfeda**

### CODE REVIEW

File: html/js/handlebars.js [hunk 1]
```diff
@@ -20,6 +20,13 @@ Handlebars.registerHelper("bytes", function (value) {
     return `${f} ${units[i]}`;
 });
 
+Handlebars.registerHelper("formatInteger", function (value) {
```
Comment (dabico) [inline html/js/handlebars.js:23]: I would rename this to `localized`

File: html/index.html [hunk 2]
```diff
@@ -588,7 +588,27 @@ <h6>Comment lines</h6>
...
+                         id="search-topics"
+                         class="form-control mb-3"
+                         placeholder="Has topic"
+                         aria-label="Repository Topic"
+                  />
+                </div>
+              </div>
+            </div>
+          </div>
+          <div class="col-12 col-md-2">
```
Comment (dabico) [inline html/index.html:611]: As stated in the meeting, move this to the top of the form, near the license and label filter. The order should be 'license', 'topic', 'label'

File: src/main/java/usi/si/seart/model/Topic.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    @OneToMany(mappedBy = "language")
+    @Fetch(value = FetchMode.JOIN)
+    Set<GitRepoMetric> metrics = new HashSet<>();
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj)
+            return true;
+        if (obj == null || getClass() != obj.getClass())
+            return false;
```
Comment (dabico) [inline src/main/java/usi/si/seart/model/Topic.java:41]: Be consistent! `GitRepoTopicKey` uses curly braces in the ifs!

File: src/main/java/usi/si/seart/service/GitRepoTopicsService.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        TopicRepository topicRepository;
+        GitRepoTopicRepository gitRepoTopicRepository;
+
+
+        @Override
+        public List<String> getAllTopicsSortByPopularity() {
+            return topicRepository.findAllSortByPopularity().stream()
+                    .map(Topic::getLabel)
+                    .limit(3000)
+                    .collect(Collectors.toList());
```
Comment (dabico) [inline src/main/java/usi/si/seart/service/GitRepoTopicsService.java:51]: This is not how it's done. All results would be loaded in memory regardless. It's better to define the repository method that excepts a `Pageable` and then define the query limit through it!

File: src/main/resources/db/migration/V1__analysis.sql [hunk 5]
```diff
@@ -18,8 +18,8 @@ CREATE TABLE repo_metrics
 
     CONSTRAINT unique_composite_key
         UNIQUE (repo_id, metric_language_id),
-    FOREIGN KEY (repo_id) REFERENCES repo(id),
-    FOREIGN KEY (metric_language_id) REFERENCES metric_language(id),
+    FOREIGN KEY (repo_id) REFERENCES repo(id) ON DELETE CASCADE,
+    FOREIGN KEY (metric_language_id) REFERENCES metric_language(id) ON DELETE CASCADE,
```
Comment (dabico) [inline src/main/resources/db/migration/V1__analysis.sql:22]: Be careful when cascading deletes across many-to-many relationships. If two rows of the left table both point to the same row in the right table, then cascading will delete the right entry if either one of the left entries is deleted. I would refrain from using cascading altogether. Better to be left with a reusable orphan than deleting more than is needed.

Comment (dabico) [reply src/main/resources/db/migration/V1__analysis.sql:22]: Also, DO NOT EDIT ALREADY EXECUTED MIGRATION FILES!

File: src/main/java/usi/si/seart/repository/TopicRepository.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+import java.util.List;
+import java.util.Optional;
+
+public interface TopicRepository extends JpaRepository<Topic, Long> {
+
+    @Query(value = "SELECT t.* FROM topics t " +
+            "INNER JOIN repo_topics grt ON t.id = grt.topic_id " +
+            "GROUP BY t.id " +
+            "ORDER BY COUNT(*) DESC", nativeQuery = true)
```
Comment (dabico) [inline src/main/java/usi/si/seart/repository/TopicRepository.java:15]: Try to refrain from using native queries. Rewrite this to use JPQL.

File: src/main/java/usi/si/seart/model/GitRepoTopicKey.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * Composite key for the repo tag.
+ */
+@Embeddable
+@AllArgsConstructor
+@NoArgsConstructor
+@Setter
+@Getter
+public class GitRepoTopicKey implements Serializable {
```
Comment (dabico) [inline src/main/java/usi/si/seart/model/GitRepoTopicKey.java:21]: nit: missing newline

**Review by dabico — CHANGES_REQUESTED** [review]
See comments

File: html/js/handlebars.js (hunk 1 above)
Comment (AlbertCerfeda) [reply html/js/handlebars.js:23]: ok boss

File: src/main/resources/db/migration/V1__analysis.sql (hunk 5 above)
Comment (AlbertCerfeda) [reply src/main/resources/db/migration/V1__analysis.sql:22]: loud and clear

File: src/main/java/usi/si/seart/service/GitRepoTopicsService.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        @Override
+        public Topic getOrCreateTopic(@NotNull String label) {
+            return topicRepository.findByLabel(label).orElseGet(() -> {
+                // If not, creates it.
+                Topic repoTopic = topicRepository.save(
+                                Topic.builder()
+                                        .label(label)
+                                        .build());
+                entityManager.clear();
+                entityManagerFactory.getCache().evict(Topic.class, repoTopic.getId());
```
Comment (dabico) [inline src/main/java/usi/si/seart/service/GitRepoTopicsService.java:61]: Are you 100% sure this is needed?

File: src/main/java/usi/si/seart/service/GitRepoTopicsService.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        public List<String> getAllTopicsSortByPopularity() {
+            return topicRepository.findAllSortByPopularity(PageRequest.of(0,3000)).stream()
+                    .map(Topic::getLabel)
+                    .collect(Collectors.toList());
+        }
+
+        @Override
+        public Topic getOrCreateTopic(@NotNull String label) {
+            return topicRepository.findByLabel(label).orElseGet(() -> {
+                // If not, creates it.
```
Comment (dabico) [inline src/main/java/usi/si/seart/service/GitRepoTopicsService.java:55]: Remove this comment

File: src/main/java/usi/si/seart/model/GitRepoTopic.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+    @ManyToOne(cascade = CascadeType.ALL)
+    @MapsId("topicId")
+    @JoinColumn(name = "topic_id")
+    Topic topic;
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj)
+            return true;
```
Comment (dabico) [inline src/main/java/usi/si/seart/model/GitRepoTopic.java:40]: nit: missing curly braces

File: src/main/java/usi/si/seart/model/GitRepoTopic.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
+package usi.si.seart.model;
+
+import lombok.*;
+import lombok.experimental.FieldDefaults;
+
+import javax.persistence.*;
```
Comment (dabico) [inline src/main/java/usi/si/seart/model/GitRepoTopic.java:6]: Star imports grrrr

File: src/main/java/usi/si/seart/service/GitRepoTopicsService.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    @PersistenceContext
+    class GitRepoTopicsServiceImpl implements GitRepoTopicsService {
+
+        TopicRepository topicRepository;
+
+        GitRepoTopicRepository gitRepoTopicRepository;
+
+        @Override
+        public List<String> getAllTopicsSortByPopularity() {
+            return topicRepository.findAllSortByPopularity(PageRequest.of(0,3000)).stream()
```
Comment (dabico) [inline src/main/java/usi/si/seart/service/GitRepoTopicsService.java:42]: I would decrease this number to the top 500, to match the labels

**Review by dabico — CHANGES_REQUESTED** [review]
Address the comments

File: src/main/java/usi/si/seart/service/GitRepoTopicsService.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    List<String> getAllTopicsSortByPopularity();
+
+    Topic getOrCreateTopic(@NotNull String languageName);
+
+    void createOrUpdateGitRepoTopics(@NotNull GitRepo repo, @NotNull List<GitRepoTopic> topics);
+
+    @Service
+    @FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
+    @AllArgsConstructor(onConstructor_ = @Autowired)
+    @PersistenceContext
```
Comment (dabico) [inline src/main/java/usi/si/seart/service/GitRepoTopicsService.java:33]: Why do we have this annotation here? I know you based this implementation on the analytics service, but here we have no `EntityManager` or associated factory, so I think it's not needed.

**Review by dabico — CHANGES_REQUESTED** [review]

**Comment by dabico** [comment]
Also, could you disable wildcard imports on the editor level? Yes, I know this can be set on the project level, but do you want to do that for every project?
Screenshot 2023-05-19 at 09 46 44

**Review by dabico — APPROVED** [review]

**Pull request merged**